src/walker.ts

```diff
@@ -0,0 +1,99 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+import { cloneDeep } from 'lodash';
+import { unwrapSteps } from './unwrapSteps';
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Promise<Pipeline>;
+  stepFn?: (step: Step) => Promise<Step>;
+  commandFn?: (command: Command) => Promise<Command>;
+}
+
+/**
+ * @param p The pipeline to evaluate
+ * @returns An evaluated pipeline with all conditionals generated if they were accepted
+ */
+export async function evaluatePipeline(p: Pipeline): Promise<Pipeline> {
+  const conditionalCache = new Map<any, any>();
+  p.steps = await unwrapSteps(p.steps, conditionalCache);
+  return p;
+}
+
+export async function walk(p: Pipeline, mutator: Mutators): Promise<Pipeline> {
+  const nodeCache = new Map<any, any>();
+
+  async function walkPipeline(p: Pipeline): Promise<Pipeline> {
+    const newSteps = [];
+    for (const step of p.steps) {
+      const result = await walkStep(step);
+      if (result) {
+        newSteps.push(result);
+      }
+    }
+    p.steps = p.steps;
+    if (mutator.pipelineFn) {
+      p = await mutator.pipelineFn(p);
+    }
+    return p;
+  }
+
+  async function walkStep(step: PotentialStep): Promise<Step> {
+    if (step instanceof Conditional) {
+      throw new Error(
+        `encountered conditional during walk, please run evaluatePipeline`,
+      );
+    }
+
+    if (nodeCache.has(step.key)) {
+      return nodeCache.get(step.key);
+    }
+    if (step instanceof CommandStep) {
+      const newDeps: Set<PotentialStep> = new Set();
+      for (const dep of step.dependencies) {
+        if (dep instanceof Conditional) {
+          throw new Error(
+            `encountered conditional during walk, please run evaluatePipeline`,
+          );
+        }
+        const newDep = await walkStep(dep);
+        if (newDep) {
+          newDeps.add(newDep);
+        }
+      }
+      step.dependencies = newDeps;
+
+      const newCommands: Command[] = [];
+      for (const command of step.command) {
+        let newCommand = cloneDeep(command);
+        newCommand = await walkCommand(newCommand);
+        newCommands.push(newCommand);
+      }
+      step.command = newCommands;
+    }
+    if (mutator.stepFn) {
+      step = await mutator.stepFn(step);
+    }
+    nodeCache.set(step.key, step);
+    return step;
+  }
+
+  async function walkCommand(command: Command): Promise<Command> {
+    const commandStr = JSON.stringify(command);
```

Comment:
You might as well use the command as the key here too if it works for the steps.

Reply:
commands don't have keys

Reply:
No I mean use the command object as the key in the map.

Reply:
Oh, that didn't work (for reasons I haven't investigated either) (as in the map didn't pickup already visited nodes). Stringifying it first seemed to fix the issue